packages/dds/tree/src/changeset/delta.ts

```diff
@@ -0,0 +1,281 @@
+ * apply the relevant `MoveIn` marks.
+ *
+ * 3. Modifications to subtrees that are also being deleted or moved are represented within the marks that describe such
+ * deletions or movements. This makes it possible specialize the type of the modifications in order to constrain the
+ * kinds of marks that can appear below them.
+ *
+ * If modify marks were not specialized then it would be possible to represent meaningless cases and consumers of this
+ * format would have to either provide implementations for them or detect when they they occur. By specializing the
+ * types we move this "detection" to the Typescript compiler.
+ *
+ * 4. Modifications of deleted and moved-out nodes are represented using modify marks within `Delete` and `MoveOut`
+ * marks. This is in opposition to a structure where the fact that a modified node is being deleted or moved-out would
+ * be represented a `Modify` mark like so:
+ * ```typescript
+ * interface ModifyAndDelete {
+ *   [type]: typeof MarkType.ModifyAndDelete;
+ *   [key: FieldKey]: (Offset | ModifyDel | MoveOut)[];
+ * }
+ * interface ModifyAndMoveOut {
+ *   [type]: typeof MarkType.ModifyAndMoveOut;
+ *   [moveId]: MoveId;
+ *   [key: FieldKey]: (Offset | ModifyOut | MoveOut)[];
+ * }
+ * export interface Delete {
+ *   [type]: typeof MarkType.Delete;
+ *   count: number;
+ * }
+ * export interface MoveOut {
+ *   [type]: typeof MarkType.MoveOut;
+ *   count: number;
+ *   moveId: MoveId;
+ * }
+ * ```
+ * Note the absence of modify information in `Delete` and `MoveOut` above.
+ *
+ * The benefit of the chosen representation over the alternative are two-fold:
+ * - It leads to less splitting of moved and deleted ranges of nodes.
+ * - It makes the format more uniform since modifications to moved-in subtrees must be represented with modifications
+ * marks within a `MoveIn` mark.
+ *
+ * 5. `MoveIn` marks are represented in the location where the content being moved resides in the input context that
+ * the delta is applied to, and `MoveOut` marks are represented in the location where the content being moved should
+ * reside after the delta is applied.
+ *
+ * The alternative would be to allow such marks to appear in temporary locations (e.g., in field "bar" for a
+ * transaction that moves content from "foo" to "bar" then moves that same content from "bar" to "baz"). This
+ * makes the format less terse and harder to reason about.
+ *
+ * 6. MoveIn marks are not inlined within `ProtoFields`.
+ *
+ * Inlining them would force the consuming code to detect `MoveIn` marks within the `ProtoFields` and handle them
+ * within the context of the insert. This would be cumbersome because either the code that is responsible for consuming
+ * the `ProtoFields` would need to be aware of and have the context to handle `MoveIn`, or some caller of that code
+ * would need to find and extract such `MoveIn` marks ahead to calling that code.
+ */
+export namespace Delta {
```

Comment:
files are modules which are namespaces. No need to put a Delta in your delta.

That said, if you want to use names that do not make sense when used unqualified, you might need to adopt some special pattern in the index.ts file when you reexport them.

Are you doing this because of issues with `export * as Delta from "./delta"` not working with API Extractor? I remember having issues with that in the past.

Anyway, our general pattern within fluid seems to be to pick names which make sense in the scope of their package, even if that means the names are a bit redundant/long in the individual files. Might be better to just match that.

Reply:
I forgot you could do `export * as Delta from "./delta"`. That seems to work and is my preference.